### PR TITLE
Fix issue where we were not offering 'use collection expression' in a nullable scenario.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/71522

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs [hunk 1]
```diff
@@ -376,10 +379,8 @@ protected bool SymbolsAreCompatible(ISymbol? symbol, ISymbol? newSymbol, bool re
                 return CompareAcrossSemanticModels(symbol, newSymbol);
             }
 
-            if (symbol.Equals(newSymbol, SymbolEqualityComparer.IncludeNullability))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:379]: @RikkiGibson @cston this is the IDE change i had to make to workaround the issue.

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs [hunk 2]
```diff
@@ -4472,4 +4473,137 @@ IEnumerable<int> M()
                 """
         }.RunAsync();
     }
+
+    [Theory, CombinatorialData, WorkItem("https://github.com/dotnet/roslyn/issues/71522")]
+    public async Task TestTargetTypedConditional1(
+        [CombinatorialValues("", "#nullable enable")] string nullable)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs:4479]: these tests demonstrate the problem.

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal

**Review by akhera99 — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson do you want me to hold off on this?  I can also merge in, and you can adjust things back if you end up changing the compiler side.

**Comment by RikkiGibson** [comment]
I have no strong opinion here about whether to merge the PR. It will maybe help buy the compiler side time to fix up the public API behavior.

I don't have a strong understanding of the code that was changed in this PR. But I thought I would point out that the compiler sometimes internally uses [TypeCompareKind.ObliviousNullableModifierMatchesAny](https://sourceroslyn.io/#Microsoft.CodeAnalysis/Symbols/TypeCompareKind.cs,d908ec8fc24bd917,references). There are times where it's acceptable for *either* an annotated or non-annotated type to unify with an oblivious type. It's possible that such a behavior is also desirable here, I'm not sure.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson ok.  i'll merge in.  you'd be able to repro the issue by keeping the tests added in this pr, but undoing the other code changes.  You'll see a call to arraysymbol's .Equals override.

**Pull request merged**